apps/settings/js/datasync/history_adapter.js

```diff
@@ -0,0 +1,53 @@
+'use strict';
+/* globals IAC */
+/* exported SynctoHistoryAdapter */
+
+var HistoryAPI = {
+  addPlaces: function(places) {
+    return IAC.request('sync-history', {
+      method: 'addPlaces',
+      args: [places]
+    });
+  },
+
+  addPlace: function(place) {
+    return IAC.request('sync-history', {
+      method: 'addPlace',
+      args: [place]
+    });
+  }
+};
+
+var SynctoHistoryAdapter = {
+  update: function(kintoCollection) {
+    function updateHistoryCollection(list) {
+      var historyRecords = list.data;
+      var places = [];
+      historyRecords.forEach(function (decryptedRecord){
+        var record = decryptedRecord.payload;
+        if(!record.histUri || !record.visits || !record.visits[0]){
+          return ;
+        }
+
+        var visits = [];
+        record.visits.forEach(function (elem){
+          visits.push(elem.date);
+        });
+
+        var place = {
+          url: record.histUri,
+          title: record.title,
+          visits: visits,
+          last_modified: decryptedRecord.last_modified
+        };
+        places.push(place);
+      });
+
+      return HistoryAPI.addPlaces(places);
```

Comment:
It would make more sense to me to have an IAC method to retrieve a sorted list of places that have been visited at least once since a certain date (i.e., since the last time we knew the data was in sync). You can do the same on the Kinto collection probably, using [filtering and sorting](http://kintojs.readthedocs.org/en/latest/api/#listing-records).

Then you can iterate over these two arrays side-by-side, updating items in both directions. When you reach the timestamp of last sync in both arrays, you can stop, because you know history was not changed retroactively. :)